src/org/omegat/core/machinetranslators/YandexCloudTranslate.java

```diff
@@ -0,0 +1,245 @@
+
+    private String AIMErrorMessage = null;
+    private String cachedAIMToken = null;
+    private long lastAIMTokenTime = 0;
+
+    @SuppressWarnings("unchecked")
+    private String getAIMToken(final String oAuthToken) {
+        if (System.currentTimeMillis() - lastAIMTokenTime > AIM_TOKEN_TTL_SECONDS * 1000) {
+
+            String request = "{\"yandexPassportOauthToken\":\"" + oAuthToken + "\"}";
+            String response;
+            Map<String, Object> rootNode;
+
+            try {
+                response = WikiGet.postJSON(IAM_TOKEN_URL, request, null);
+            } catch (WikiGet.ResponseError e) {
+                // Try to extract error message from the error body
+                AIMErrorMessage = extractErrorMessage(e.body);
+                if (AIMErrorMessage == null) {
+                    AIMErrorMessage = OStrings.getString("MT_ENGINE_YANDEX_CLOUD_BAD_AIM_RESPONSE");
+                }
+                return null;
+            } catch (IOException e) {
+                AIMErrorMessage = e.getLocalizedMessage();
+                return null;
+            }
+
+            try {
+                rootNode = (Map<String, Object>) JsonParser.parse(response);
+                cachedAIMToken = (String) rootNode.get("iamToken");
+                lastAIMTokenTime = System.currentTimeMillis();
+            } catch (Exception e) {
+                Log.logErrorRB(e, "MT_ENGINE_YANDEX_CLOUD_BAD_AIM_RESPONSE");
+                AIMErrorMessage = OStrings.getString("MT_ENGINE_YANDEX_CLOUD_BAD_AIM_RESPONSE");
+                return null;
+            }
+        }
+        return cachedAIMToken;
+    }
+
+    @Override
+    protected String getPreferenceName() {
+        return Preferences.ALLOW_YANDEX_CLOUD_TRANSLATE;
+    }
+
+    @Override
+    public String getName() {
+        return OStrings.getString("MT_ENGINE_YANDEX_CLOUD");
+    }
+
+    @Override
+    protected String translate(Language sLang, Language tLang, String text) {
+        String trText = text.length() > MAX_TEXT_LENGTH ? text.substring(0, MAX_TEXT_LENGTH - 3) + "..." : text;
+        String prev = getFromCache(sLang, tLang, trText);
+        if (prev != null) {
+            return prev;
+        }
+
+        String oAuthToken = getCredential(PROPERTY_OATH_TOKEN);
+        if (oAuthToken == null || oAuthToken.isEmpty()) {
+            return OStrings.getString("MT_ENGINE_YANDEX_CLOUD_OAUTH_TOKEN_NOT_FOUND");
+        }
+
+        String folderId = getCredential(PROPERTY_FOLDER_ID);
+        if (folderId != null && !folderId.isEmpty()) {
+            return OStrings.getString("MT_ENGINE_YANDEX_CLOUD_FOLDER_ID_NOT_FOUND");
+        }
+
+        String AIMToken = getAIMToken(oAuthToken);
+        if (AIMToken == null) {
+            return AIMErrorMessage;
+        }
+
+        StringBuilder requestBuilder = new StringBuilder();
+        requestBuilder.append("{");
```

Comment:
You can combine a lot of these calls in a "fluent" fashion. It might help readability to group related pieces together.